src/rhash/hash.c

```diff
@@ -1143,6 +1143,92 @@ static int rc_hash_nintendo_ds(char hash[33], const char* path)
   return rc_hash_finalize(&md5, hash);
 }
 
+static int rc_hash_gamecube(char hash[33], const char* path)
+{
+  md5_state_t md5;
+  void* file_handle;
+  const uint32_t BASE_HEADER_SIZE = 0x2440;
+
+  file_handle = rc_file_open(path);
+  uint8_t* buffer = (uint8_t*)malloc(MAX_BUFFER_SIZE);
+  if (!buffer)
+  {
+    rc_file_close(file_handle);
+    return rc_hash_error("Could not allocate temporary buffer");
+  }
+
+  // Verify Gamecube
+  rc_file_seek(file_handle, 0, SEEK_SET);
+  rc_file_read(file_handle, buffer, 0x2500);
+  if (buffer[0x1c] != 0xC2|| buffer[0x1d] != 0x33 || buffer[0x1e] != 0x9F || buffer[0x1f] != 0x3D)
+  {
+    rc_file_close(file_handle);
+    free(buffer);
+    return rc_hash_error("Not a Gamecube disc");
+  }
+
+  // GetApploaderSize
+  uint32_t apploader_header_size = 0x20;
+  uint32_t apploader_body_size = 
+    (buffer[BASE_HEADER_SIZE + 0x14] << 24) |
+    (buffer[BASE_HEADER_SIZE + 0x15] << 16) |
+    (buffer[BASE_HEADER_SIZE + 0x16] << 8) |
+    buffer[BASE_HEADER_SIZE + 0x17];
+  uint32_t apploader_trailer_size =
+    (buffer[BASE_HEADER_SIZE + 0x18] << 24) |
+    (buffer[BASE_HEADER_SIZE + 0x19] << 16) |
+    (buffer[BASE_HEADER_SIZE + 0x1a] << 8) |
+    buffer[BASE_HEADER_SIZE + 0x1b];
+  uint32_t header_size = BASE_HEADER_SIZE + apploader_header_size + apploader_body_size + apploader_trailer_size;
+
+  // Hash headers
+  md5_init(&md5);
+  md5_append(&md5, buffer, header_size);
+  if (verbose_message_callback)
+  {
+    char message[128];
+    snprintf(message, sizeof(message), "Hashing %u byte header", header_size);
+    verbose_message_callback(message);
+  }
+
+  // GetBootDOLOffset
+  uint32_t dol_offset = (buffer[0x420] << 24) | (buffer[0x421] << 16) | (buffer[0x422] << 8) | buffer[0x423];
+
+  // Iterate through the 7 main.dol code segments
+  rc_file_seek(file_handle, dol_offset, SEEK_SET);
+  rc_file_read(file_handle, buffer, MAX_BUFFER_SIZE);
+  for (size_t i = 0; i < 7; i++)
+  {
+    uint32_t offset = (buffer[0x0 + i * 4] << 24) | (buffer[0x1 + i * 4] << 16) | (buffer[0x2 + i * 4] << 8) | buffer[0x3 + i * 4];
+    uint32_t size = (buffer[0x90 + i * 4] << 24) | (buffer[0x91 + i * 4] << 16) | (buffer[0x92 + i * 4] << 8) | buffer[0x93 + i * 4];
```

Comment:
I noticed that several code segments had 0 size. You can skip the hash/log messages for those. You could also skip the seek/read if you follow my suggestion to not preload 64MB of data.